src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/SpatioTemporalZonalStatistics.java

```diff
@@ -0,0 +1,583 @@
+                        features.add(builder.buildFeature(zone.getID()));
+                    }
+                } catch (Exception e) {
+                    throw new ProcessException("Failed to compute statistics on feature " + zone, e);
+                }
+            }
+            // return the first feature in the current buffer
+            SimpleFeature f = features.remove(0);
+            return f;
+        }
+
+        /** Add the statistics to the feature builder */
+        private void addStatsToFeature(StatisticsAggregator stats, Set<Statistic> requestedStats) {
+            double count = stats.getAggregatedCount();
+            builder.add(count); // count
+            addDynamicStatsToFeature(builder, stats, requestedStats);
+        }
+
+        private void addDynamicStatsToFeature(
+                SimpleFeatureBuilder builder, StatisticsAggregator stats, Set<Statistic> requestedStats) {
+            if (requestedStats.contains(Statistic.MIN)) {
+                builder.add(stats.getAggregatedMin());
+            }
+            if (requestedStats.contains(Statistic.MAX)) {
+                builder.add(stats.getAggregatedMax());
+            }
+            if (requestedStats.contains(Statistic.SUM)) {
+                builder.add(stats.getAggregatedSum());
+            }
+            if (requestedStats.contains(Statistic.MEAN)) {
+                builder.add(stats.getAggregatedMean());
+            }
+            if (requestedStats.contains(Statistic.MEDIAN)) {
+                builder.add(stats.getAggregatedMedian());
+            }
+        }
+
+        private StatisticsAggregator processStatistics(Geometry geometry) throws TransformException, IOException {
+            GridCoverage2D refCoverage = null;
+            GridCoverage2D dataCoverage = null;
+            GridCoverage2D cropped = null;
+            ReferencedEnvelope geometryEnvelope = null;
+            List<Range<Double>> noDataValueRangeList = null;
+            ROI roi = null;
+            ParameterValue<List> timeParam = AbstractGridFormat.TIME.createValue();
+            Statistic[] reqStatsArr = requestedStats.toArray(new Statistic[0]);
+            StatisticsAggregator aggregator = new StatisticsAggregator(requestedStats);
+
+            for (Object temporalItem : times) {
+                if (temporalItem instanceof Date) {
+                    Date time = (Date) temporalItem;
+                    timeParam.setValue(Collections.singletonList(time));
+                } else if (temporalItem instanceof DateRange) {
+                    DateRange range = (DateRange) temporalItem;
+                    timeParam.setValue(Collections.singletonList(range));
+                } else {
+                    throw new IllegalArgumentException("Unsupported temporal item: " + temporalItem);
+                }
+                GeneralParameterValue[] gpv = {timeParam};
+                try {
+                    dataCoverage = reader.read(gpv);
```

Comment:
This might be dangerous, if the reader can only do immediate read, I would suggest trying to read only the portion we need (get the bbox of the current zone, make sure partial pixels are covered just in case by increasing it by the pixel resolutions in all directions)

Reply:
Fixed